README.md
[![LdapSaisie](src/images/default/logo.png)](https://ldapsaisie.org)

LdapSaisie is a Web application developed to manage LDAP directory. It has been written in PHP / JavaScript and is published under the GNU GPL license. This application is designed to abstract the complexity of the directory through a simple and intuitive administration interface. It was designed with the objective of maximal modularity and easy extension or adaptation: you can use modules, extensions and plugins. LdapSaisie allows any system administrator to manage data stored inside a LDAP server, and thus administrate its information system in a simple manner. It's also possible to let users access their own data, with read-only or read-write access.

__[Project website](https://ldapsaisie.org)__ | __[Documentation](https://ldapsaisie.org/doc/full)__ | __[Demonstration](https://ldapsaisie.org/demo)__

## Key features

* Management of one or several directories;
* Management of simple and multi-branch directories;
* Able to manage unlimited objects types;
* Allow unlimited users to connect to the interface;
* A smooth rights management allows to manage each object and attributes access rights, and allows to delegate rights;
* Able to manage a lot of attribute type. Each attribute is given specific features which ease application navigation (Automatic password generation, field auto-completion);
* Able to manage high rate of attributes verification rules;
* Easy management of objects relationships;
* Easy modification of application design using templates;
* Management of personalized triggers, which allows to execute you own scripts, functions or methods at any action. The triggers scope is set to able a whole modification of application behavior;
* Smooth attributes visualization management, allowing to auto-modify attributes according to others;
* Possibility to manage hidden attributes.
* Customizable responsive web interface

## Screenshots

Please see [project website](https://ldapsaisie.org/en:screenshot) for some screenshots of the web interface. Keep in mind that the web interface could be personalized to feet with your needs.

## Demonstration

A demonstration version is available at :

http://ldapsaisie.org/demo

This demonstration interface is connected to a sample LDAP directory such that the schema and the imported LDIF file are provided with the sources of the application (in the `lsexample` directory).

Several demonstration accounts exist and have more or less extensive rights. These different accounts are presented on the login page.

## Installation

### Requirements

* PHP (>= 5.6)
* PEAR [NetLDAP2](http://pear.php.net/package/Net_LDAP2) package
* [Smarty](http://www.smarty.net/)

Some others dependencies exists for specific features, please see [documentation](https://ldapsaisie.org/doc/all-in-one/LdapSaisie.html#install-requirements) for details (french).

### Download

#### Using Debian packages

You can easily install LdapSaisie using Debian packages by using the following command to configure the project's repository :
```
echo "deb http://ldapsaisie.org/debian buster main" | tee /etc/apt/sources.list.d/ldapsaisie.list
wget -O - http://ldapsaisie.org/debian/ldapsaisie.gpg.key | apt-key add -
apt-get update
apt-get install ldapsaisie ldapsaisie-archive-keyring
```

__Note:__ Hourly generated Debian packages from _master_ branch are also provided in the _sid_ Debian repository of the project.  To use these packages, use the following command to configure the project's _sid_ repository:

```
echo "deb http://ldapsaisie.org/debian sid main" | tee /etc/apt/sources.list.d/ldapsaisie.list
```

#### Using Git

The Git repos can be cloned anonymously with the command :

```
git clone https://gitlab.easter-eggs.com/ee/ldapsaisie.git
```

#### Snapshot

Every 15 minutes, a snapshot of the Git repository is built and available for download here:

http://ldapsaisie.org/download/ldapsaisie-snapshoot.tar.gz

### Configuration

It's the main step of the installation process and the most complicated. Firstly, you have to configure the global configuration mostly contain in the file `conf/config.inc.php`. Second, you will have to configure your object types and their relationships. You could refer to example files provided with the application and to the [official documentation](https://ldapsaisie.org/doc/all-in-one/LdapSaisie.html#config-LSobject) (in french).

__Note:__ A more details [tutorial](https://ldapsaisie.org/doc/all-in-one/LdapSaisie.html#install-tutorial) (in french) is also provided in official documentation.

## Upgrade

### Using Debian packages

When using Debian packages, the upgrading process to pretty simple: just upgrade the package:

```
apt-get update
apt-get install ldapsaisie
```

Once the application has been updated, paid attention to new features and points of vigilance described in [official documentation](https://ldapsaisie.org/doc/all-in-one/LdapSaisie.html#upgrade) (in french).

### Using Git

When using git, you could use the `upgradeFromGit.sh` which automates the update if you have followed the installation procedure for it.

This script will then take care of:

* Clean the working-tree Git from the symbolic links of the local files (and possibly the theme) set up during a previous execution;
* Empty the template cache;
* Update the Git working-tree via a `git pull`;
* Install symbolic links for local files. In case of locally modified files, the script will notify you and will allow you to simply update your local file (via a `vim -d`);
* Detect changes in the MO files (translation) and in this case trigger a reload of the web server to be taken into account;
* Option: to compile an up-to-date local version of the documentation;

Once the application has been updated, paid attention to new features and points of vigilance described in [official documentation](https://ldapsaisie.org/doc/all-in-one/LdapSaisie.html#upgrade) (in french).

## License

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program (see `LICENSE` file); if not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.


debian/changelog
ldapsaisie (4.0.0-1) unstable; urgency=medium

  * Add API feature
  * Add full import / export objets feature (see LSio & LSioFormat)
  * Global improvements on SUPANN support:
    * nomenclatures:
      * Update/clean nomenclatures
      * Normalize affiliations & civilites nomenclatures
      * rework on nomenclatures loading
    * Add support for new attributes :
      * supannMailPrive
      * supannAdressePostalePrivee
      * supannTelephonePrive
      * supannOIDCGenre
      * supannCodePopulation
      * supannEmpProfil
      * supannExtProfil
      * supannEtuDiplome
      * supannEtuElementPedagogique
      * supannEtuEtape
      * supannEtuRegimeInscription
      * supannEtuSecteurDisciplinaire
      * supannEtuTypeDiplome
      * supannRessourceEtat
      * supannRessourceEtatDate
    * add new components of supannEtuInscription attribute (SUPANN 2021)
    * add generate_supann_codeEtablissement_uai_nomenclature CLI command
    * add eduPersonUniqueId()
    * supannParseCompositeValue(): check global value integrity
    * Improve SUPANN attributes CSS
  * Add dynamic groups support (with cache members attributes)
  * Ensure attributes values are always an array in LSattribute & LSattr_ldap
  * Remove historic read right on all attributes of the hard-coded "admin" LSprofile
  * LSsession:
    * improve loadLSaddons() method
    * fix errors in changeAuthUser() and add logging messages
    * Fix/improve LSsession_13 errors (unconfigured/unknown custom action function)
    * checkAccess(): fix checking read right
    * Fix relationCanAccess() method and add debuging infos
    * displayAjaxReturn(): Adjust HTTP returned code on unsuccessfull request
    * getLSuserObject(): improve errors logging
    * setApiMode now trigger LSsession :: setAjaxDisplay()
  * LSldap: fix search & getNumberResult methods to use basedn from configuration instead of empty
    one.
  * LSldapObject:
    * make method _updateData() protected instead of private
    * fix/improve validateAttrData()
    * submitChange():
      * on new object, check if generated DN is not already affected
      * Add security check about isNew() return
      * fix error renaming LSobject
    * fix default $attrValues handling for updateObjectsInRelation/deleteOneObjectInRelation methods
    * clean update/validate form data code to fix handling just check mode
    * fix triggering after_modify event on create
    * getValue(): ensure return is array in case of attribute
  * LSrelation: add display_name_format parameter
  * LSattribute:
    * fix canBeGenerated() when default_value isn't a string
    * fix handling default_value in generateValue()
    * add default LDAP & HTML type (ascii / text)
    * addToForm(): fix checking user rights before adding/freezing it
    * fix handling current/provided data
    * addToForm(): fix freezing
  * LSattr_ldap:
    * shadowExpire:
      * fix getUpdateData()
      * fix to rounding method to avoid date change
    * compositeValueToJSON: use static instead of self keyword (to allow override)
    * Add pwdHistory
    * date: add note about date & time with micro-seconds
    * naiveDate: Add format option
    * password: fix verify method for Salted SHA2 passwords
    * sambaAcctFlagsparse_flags(): fix handling null|empty value
  * LSattr_html:
    * jsonCompositeAttribute: Add fullWidth parameter and improve responsive display
    * sambaAcctFlags: Add note about how setting default_value
    * valueWithUnit: fix handling zero value
    * select_box: add inline parameter
    * fix handling current/provided data
    * add $singleValue class variable to handle this common case
  * LSform:
    * fix displaying error in form in case of global error
    * fix displaying multiple errors on same field (on Ajax submit)
    * display fields errors in same order as triggered
    * fix reporting failure on Ajax submit
  * LSformElement:
    * getApiValue(): add $details parameter
    * getEmptyField(): do not set value to null (already set by fetchTemplate())
    * labeledValue:
      * fix/improve display
      * add translate_labels parameter and fix translating it
      * fix label translation on freeze view
    * date: fix getPostData() method
    * mail:
      * Move autocomplete feature in text parent class
      * fix duplicated email JS button
      * supannLabeledValue: Add validation of POST values from configured nomenclatures
    * image: fix bug deleting current image
    * password:
      * fix using LSldapobject getDisplayFData method instead of getFData
      * fix sendind email in create form
      * fix multiple sent email in some cases
    * supannLabeledValue:
      * fix unparsable value display
      * improve API mode
    * supannCompositeAttribute:
      * Add possibility to override parse/format value methods to handledifferent composite value
      formats
      * Add possibility to configure date components LDAP format
      * handle API/CLI mode
  * LSformRule:
    * add validate_values() method and use it to value rules
    * do not defined validate_one_by_one constant visibility
    * add possibility to throw custom exception to provide error details
    * Improve LSformRules to throw LSformRuleException with error details and only show default
      error message if rule does not trigger itself an error
    * Handle rule without options or message
    * Add numberOfValues rule
    * configure CLI parameters autocompleters
    * compare: code cleaning
    * add zxcvbn rule for password security check
  * LSsearch:
    * rename method setParamsFormPostData() to setParamsFromRequest()
    * setParamsFromRequest(): add check on parameters load from request data
    * setParamsFromRequest(): fix error handling filter parameter
  * LSaddons:
    * mail: Add test_send_mail CLI command
    * mailquota: improve error reporting
    * posix: fix generate_shadowLastChange()
    * samba: fix PHP warning about LS_SAMBA_UNIX_ID_POOL_DN contant
    * LSaccessRightsMatrixView:

trunk/view.php
<?php
/*******************************************************************************
 * Copyright (C) 2007 Easter-eggs
 * http://ldapsaisie.labs.libre-entreprise.org
 *
 * Author: See AUTHORS file in top-level directory.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

******************************************************************************/

require_once 'core.php';

if(LSsession :: startLSsession()) {
  if (isset($_REQUEST['LSobject'])) {
    $LSobject = $_REQUEST['LSobject'];
    $dn = $_REQUEST['dn'];
    
    if (LSsession :: in_menu($LSobject)) {
    
      if ( $LSobject == 'SELF' ) {
        $LSobject = LSsession :: getLSuserObject() -> getType();
        $dn = LSsession :: getLSuserObjectDn();
      }
      
      if ( LSsession :: loadLSobject($LSobject) ) {
        // Affichage d'un objet
        if ( $dn!='' ) {
          if (LSsession :: canAccess($LSobject,$dn)) {
            if ( LSsession :: canEdit($LSobject,$dn) ) {
              $LSview_actions[] = array(
                'label' => _('Modify'),
                'url' =>'modify.php?LSobject='.$LSobject.'&amp;dn='.$dn,
                'action' => 'modify'
              );
            }
            
            if (LSsession :: canCreate($LSobject)) {
              $LSview_actions[] = array(
                'label' => _('Copy'),
                'url' =>'create.php?LSobject='.$LSobject.'&amp;load='.$dn,
                'action' => 'copy'
              );
            }
            
            if (LSsession :: canRemove($LSobject,$dn)) {
              $LSview_actions[] = array(
                'label' => _('Delete'),
                'url' => 'remove.php?LSobject='.$LSobject.'&amp;dn='.$dn,
                'action' => 'delete'
              );
            }
            
            if (LSsession :: getLSuserObjectDn() != $dn) {
              $object = new $LSobject();
              $object -> loadData($dn);
              $GLOBALS['Smarty'] -> assign('pagetitle',$object -> getDisplayName());
            }
            else {
              $object = LSsession :: getLSuserObject();
              $GLOBALS['Smarty'] -> assign('pagetitle',_('My account'));
            }
            
            $view = $object -> getView();
            $view -> displayView();
            
            // LSrelations
            if (LSsession :: loadLSclass('LSrelation')) {
              LSrelation :: displayInLSview($object);
            }
            
            $GLOBALS['Smarty'] -> assign('LSview_actions',$LSview_actions);
            LSsession :: setTemplate('view.tpl');
          }
          else {
            LSerror :: addErrorCode('LSsession_11');
          }
        }
        // Affichage d'une liste d'un type d'objet
        elseif (LSsession :: loadLSclass('LSsearch')) {
          $object = new $LSobject();
          $GLOBALS['Smarty']->assign('pagetitle',$object -> getLabel());
          
          $LSsearch = new LSsearch($LSobject,'LSview');
          $LSsearch -> setParamsFormPostData();
          
          $searchForm = array (
            'action' => $_SERVER['PHP_SELF'],
            'recursive' => (! LSsession :: isSubDnLSobject($LSobject) ),
            'labels' => array (
              'submit' => _('Search'),
              'approx' => _('Approximative search'),
              'recursive' => _('Recursive search')
            ),
            'values' => array (
              'pattern' => $LSsearch->getParam('pattern'),
              'approx' => $LSsearch->getParam('approx'),
              'recursive' => $LSsearch->getParam('recursive')
            ),
            'names' => array (
              'submit' => 'LSsearch_submit'
            ),
            'hiddenFields' => $LSsearch -> getHiddenFieldForm()
          );
          $GLOBALS['Smarty']->assign('searchForm',$searchForm);
          
          $LSview_actions=array();
          if(LSsession :: canCreate($LSobject)) {
            $LSview_actions['create'] = array (
              'label' => 'New',
              'url' => 'create.php?LSobject='.$LSobject,
              'action' => 'create'
            );
          }
          $LSview_actions['refresh'] = array (
            'label' => 'Refresh',
            'url' => 'view.php?LSobject='.$LSobject.'&amp;refresh',
            'action' => 'refresh'
          );
          /*$LSview_actions['purge'] = array (
            'label' => 'Purge the cache',
            'url' => 'view.php?LSobject='.$LSobject.'&amp;LSsearchPurgeSession',
            'action' => 'delete'
          );*/
          $GLOBALS['Smarty']->assign('LSview_actions',$LSview_actions);
          
          $LSsearch -> run();
          $page=(int)$_REQUEST['page'];
          $page = $LSsearch -> getPage($page);
          $GLOBALS['Smarty']->assign('page',$page);
          $GLOBALS['Smarty']->assign('LSsearch',$LSsearch);
          
          if (LSsession :: loadLSclass('LSform')) {
            LSform :: loadDependenciesDisplayView();
          }
          
          LSsession :: setTemplate('viewSearch.tpl');
        }
        else {
          LSsession :: addErrorCode('LSsession_05','LSsearch');
        }
      }
    }
    else {
      LSerror :: addErrorCode('LSsession_11');
    }
  }
  else {
    LSerror :: addErrorCode('LSsession_12');
  }
}
else {
  LSsession :: setTemplate('login.tpl');
}
// Affichage des retours d'erreurs
LSsession :: displayTemplate();

if (isset($LSsearch)) {
  $LSsearch->afterUsingResult();
}

?>
